executor/stale_txn_test.go

```diff
@@ -772,3 +772,80 @@ func (s *testStaleTxnSuite) TestSetTransactionInfoSchema(c *C) {
 	tk.MustExec("commit")
 	c.Assert(tk.Se.GetInfoSchema().SchemaMetaVersion(), Equals, schemaVer3)
 }
+
+func (s *testStaleTxnSuite) TestStaleSelect(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	defer tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (id int)")
+
+	tolerance := 50 * time.Millisecond
+
+	tk.MustExec("insert into t values (1)")
+	time.Sleep(tolerance)
+	time1 := time.Now()
+
+	tk.MustExec("insert into t values (2)")
+	time.Sleep(tolerance)
+	time2 := time.Now()
+
+	tk.MustExec("insert into t values (3)")
+	time.Sleep(tolerance)
+
+	staleRows := testkit.Rows("1")
+	staleSQL := fmt.Sprintf(`select * from t as of timestamp '%s'`, time1.Format("2006-1-2 15:04:05.000"))
+
+	// test normal stale select
+	tk.MustQuery(staleSQL).Check(staleRows)
+
+	// test stale select in txn
+	tk.MustExec("begin")
+	c.Assert(tk.ExecToErr(staleSQL), NotNil)
+	tk.MustExec("commit")
+
+	// test prepared stale select
+	tk.MustExec(fmt.Sprintf(`prepare s from "%s"`, staleSQL))
+	tk.MustQuery("execute s")
+
+	// test prepared stale select in txn
+	tk.MustExec("begin")
+	tk.MustQuery("execute s").Check(staleRows)
+	tk.MustExec("commit")
+
+	// test stale select in stale txn
+	tk.MustExec(fmt.Sprintf(`start transaction read only as of timestamp '%s'`, time2.Format("2006-1-2 15:04:05.000")))
+	c.Assert(tk.ExecToErr(staleSQL), NotNil)
+	tk.MustExec("commit")
+
+	// test prepared stale select in stale txn
+	tk.MustExec(fmt.Sprintf(`start transaction read only as of timestamp '%s'`, time2.Format("2006-1-2 15:04:05.000")))
+	tk.MustQuery("execute s").Check(staleRows)
+	tk.MustExec("commit")
```

Comment:
Is `execute s` expected to be executed successfully in staleness transaction while the staleSQL are not?

Reply:
How about not in prepare?

Reply:
Yes, it expected. And it should be possible to throw an error in `OptimizePreparedPlan`.

Reply:
IMO this is not expected. We should avoid multiple snapshotTS in a transaction as users may think it's data inconsistency error during transaction.